File: blender-plugin/2.8/sortblend/strid.py
#    This file is a part of SORT(Simple Open Ray Tracing), an open-source cross
#    platform physically based renderer.
#
#    Copyright (c) 2011-2019 by Jiayin Cao - All rights reserved.
#
#    SORT is a free software written for educational purpose. Anyone can distribute
#    or modify it under the the terms of the GNU General Public License Version 3 as
#    published by the Free Software Foundation. However, there is NO warranty that
#    all components are functional in a perfect manner. Without even the implied
#    warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
#    General Public License for more details.
#
#    You should have received a copy of the GNU General Public License along with
#    this program. If not, see <http://www.gnu.org/licenses/gpl-3.0.html>.

# cooresponding implementation of strid.h in python code.
# The algorithm has to be exactly the same with the one defined in strid.h

# Computation of cyclic redundancy checks
# https://en.wikipedia.org/wiki/Computation_of_cyclic_redundancy_checks
def crc32_bitwise( data , len ):
    Polynomial = 0xEDB88320

    crc = 0
    i = 0
    while i < len:
        crc ^= ord(data[i])

        crc = (crc >> 1) ^ (-int(crc & 1) & Polynomial)
        crc = (crc >> 1) ^ (-int(crc & 1) & Polynomial)
        crc = (crc >> 1) ^ (-int(crc & 1) & Polynomial)
        crc = (crc >> 1) ^ (-int(crc & 1) & Polynomial)
        crc = (crc >> 1) ^ (-int(crc & 1) & Polynomial)
        crc = (crc >> 1) ^ (-int(crc & 1) & Polynomial)
        crc = (crc >> 1) ^ (-int(crc & 1) & Polynomial)
        crc = (crc >> 1) ^ (-int(crc & 1) & Polynomial)

        i += 1

    # Technically speaking, this is a signed integer, but since the bits are the same and
    # the value is never used in a mathematic operation, this is not a big problem.
    return ~crc

# convert a string to hashed key
def SID( str ):
    return crc32_bitwise( str , len(str) )

File: src/unittests/strid.cpp
/*
    This file is a part of SORT(Simple Open Ray Tracing), an open-source cross
    platform physically based renderer.

    Copyright (c) 2011-2022 by Jiayin Cao - All rights reserved.

    SORT is a free software written for educational purpose. Anyone can distribute
    or modify it under the the terms of the GNU General Public License Version 3 as
    published by the Free Software Foundation. However, there is NO warranty that
    all components are functional in a perfect manner. Without even the implied
    warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
    General Public License for more details.

    You should have received a copy of the GNU General Public License along with
    this program. If not, see <http://www.gnu.org/licenses/gpl-3.0.html>.
*/

#include "thirdparty/gtest/gtest.h"
#include "core/strid.h"
#include "unittest_common.h"

using namespace unittest;

TEST(StringID, BasicChecking) {
    // Make sure sid generated in different ways result in the same value
    StringID    sid0( "hello world" );
    EXPECT_EQ( sid0 , SID( "hello world" ) );

    // Make sure sid generated the same ways result in the same value.
    StringID    sid1( "hello world" );
    EXPECT_EQ( sid1 , sid0 );

    // See if there is collision, passing this test mostly like means nothing,
    // but failing to pass the test is clearly wrong.
    StringID    sid2( "Hello world" );
    EXPECT_NE( sid0 , sid2 );
    EXPECT_NE( sid0 , SID( "Hello world") );
}

TEST(StringID, Invalid_StringID) {
    // Empty string should result in INVALID_SID
    EXPECT_EQ( SID("") , INVALID_SID );
}